cpp/negentropy_wrapper.c

```diff
@@ -0,0 +1,229 @@
+{
+   for (int i=0; i < from_ids.size(); i ++){    
+    to_ids[i].len = from_ids[i].size();
+    to_ids[i].data =  (unsigned char*)from_ids[i].c_str();
+   }
+}
+
+int reconcile_with_ids(void* negentropy, buffer*  query,reconcile_cbk cbk, char* outptr){
+    Negentropy<negentropy::storage::BTreeMem> *ngn_inst;
+    ngn_inst = reinterpret_cast<Negentropy<negentropy::storage::BTreeMem>*>(negentropy);
+
+    std::optional<std::string> out;
+    std::vector<std::string> haveIds, needIds;
+    uint64_t have_ids_len, need_ids_len;
+    buffer* have_ids;
+    buffer* need_ids;
+
+    try {
+        out = ngn_inst->reconcile(std::string_view(reinterpret_cast< char const* >(query->data), query->len), haveIds, needIds);
+
+        have_ids_len = haveIds.size();
+        need_ids_len = needIds.size();
+        have_ids = (buffer*)malloc(have_ids_len*sizeof(buffer));
+        need_ids = (buffer*)malloc(need_ids_len*sizeof(buffer));
+
+        std::cout << "have_ids_len:" << have_ids_len << "need_ids_len:" << need_ids_len << std::endl;
+
+        transform(haveIds, have_ids);
+        transform(needIds, need_ids);
+    } catch(negentropy::err e){
+        std::cout << "caught error "<< e.what() << std::endl;
+        //TODO:Find a way to return this error and cleanup partially allocated memory if any
+        return -1;
+    }
+    buffer output = {0,NULL};
+    if (out) {
+        output.len = out.value().size();
+        output.data = (unsigned char*)out.value().c_str();
+        std::cout << "reconcile_with_ids output of reconcile is, len:" << out.value().size() << ", output:";
+        printHexString(std::string_view(out.value()));
+    }
+    std::cout << "invoking callback" << std::endl;
+    std::flush(std::cout);
+
+    cbk(have_ids, have_ids_len, need_ids, need_ids_len, &output, outptr);
+    std::cout << "invoked callback" << std::endl;
+    std::flush(std::cout);
+
+    free(have_ids);
+    free(need_ids);
+    return 0;
+}   
+
+void transform_with_alloc(std::vector<std::string> &from_ids, buffer* to_ids)
+{
+   for (int i=0; i < from_ids.size(); i ++){
+    to_ids[i].data = (unsigned char*) malloc(from_ids[i].size()*sizeof(unsigned char));  
+    to_ids[i].len = from_ids[i].size();
+    memcpy(to_ids[i].data, from_ids[i].c_str(),to_ids[i].len);
+   }
+}
+
+void reconcile_with_ids_no_cbk(void* negentropy, buffer*  query, result* result){
+    Negentropy<negentropy::storage::BTreeMem> *ngn_inst;
+    ngn_inst = reinterpret_cast<Negentropy<negentropy::storage::BTreeMem>*>(negentropy);
+
+    std::optional<std::string> out;
+    std::vector<std::string> haveIds, needIds;
+    try {
+        out = ngn_inst->reconcile(std::string_view(reinterpret_cast< char const* >(query->data), query->len), haveIds, needIds);
+
+        result->have_ids_len = haveIds.size();
+        result->need_ids_len = needIds.size();
+        result->have_ids = (buffer*)malloc(result->have_ids_len*sizeof(buffer));
```

Comment:
Where is this memory being released?

Reply:
I had implemented a free_result function which releases memory for all these buffers.